Show data point values on static viz

WiNloSt opened the pull request:
> **Warning**
> There are a few hacks I make on BE to show the actual result, and those shouldn't be merged. Once the BE is properly implemented these hacks should be removed from this PR.
> - https://github.com/metabase/metabase/pull/24813/commits/f1ce21efbf11e6d3624eab286171e92d18a01ca6
> - https://github.com/metabase/metabase/pull/24813/commits/c5d1b1be685f74c95ddc7a07ebc313be142ebf29

EPIC: #24759
Issue: #20554

## Todos
- [x] Bar
- [x] Line
- [x] Area
- [x] Stacked Area
- [x] Long values (shorten numbers like 100k, 1m)
- [x] Crowded lines
- [x] Crowded bars
- [x] Crowded areas

## How to test
- Checkout this PR and add changes from these 2 commits manually (without them static viz won't render any data point values when rendering with BE)
    - https://github.com/metabase/metabase/commit/f1ce21efbf11e6d3624eab286171e92d18a01ca6
    - https://github.com/metabase/metabase/commit/c5d1b1be685f74c95ddc7a07ebc313be142ebf29
- Set up an email e.g. using `maildev` with `docker run -p 25:25 -p 80:80 -d --rm maildev/maildev:1.1.0`
- Start adding a bunch of charts with data point value enabled to a dashboard and send a dashboard subscription.
    - The reason to test with dashboard subscription because you can instantly send an email rather than having to wait for a pulse
    ![image](https://user-images.githubusercontent.com/1937582/186085758-62dfa7a2-11bb-4e60-bd57-a11ecb7d38dd.png)

---
The following todos will be addressed in a separate PR (https://github.com/metabase/metabase/pull/24938)
- [ ] Multiseries bars
- [ ] Close to chart edges labels should be visible
- [ ] Negative bars


<img width="640" alt="Screen Shot 2022-08-17 at 1 24 09 PM" src="https://user-images.githubusercontent.com/14301985/185084090-80224720-9523-4171-b02c-e62774deb914.png">


<img width="638" alt="Screen Shot 2022-08-17 at 1 24 17 PM" src="https://user-images.githubusercontent.com/14301985/185084128-4fe21a1e-591d-44cf-ac9a-68c71ee281fd.png">

---

## Results

### Bar chart
#### Dashboard
![image](https://user-images.githubusercontent.com/1937582/184912766-bd99b3d6-2f4d-4ab3-a281-970befe9884d.png)

#### Static viz
![image](https://user-images.githubusercontent.com/1937582/184912618-3d2098ed-373d-490e-bb9b-11f2f94f9a6a.png)

### Line chart
#### Dashboard
![image](https://user-images.githubusercontent.com/1937582/184912537-b75384b4-68df-4b65-be7c-5d288ea298db.png)

#### Static viz
![image](https://user-images.githubusercontent.com/1937582/184912556-8f0f9741-4760-461a-9136-f582822be145.png)

### Area chart
#### Dashboard
![image](https://user-images.githubusercontent.com/1937582/184912716-ac4d751e-a1c4-46e1-ba35-2ee2a83f4035.png)

#### Static viz
![image](https://user-images.githubusercontent.com/1937582/184912661-77d0a24f-e80d-402f-a57b-f7e74fe75a9c.png)

### Stacked area chart
![image](https://user-images.githubusercontent.com/1937582/185419380-18b88d0c-f31d-43c5-9333-e2b9aca31ad9.png)

### Automatic compact value format
![image](https://user-images.githubusercontent.com/1937582/185419575-c2045ad3-63f8-4e6d-9611-c9605ec1f087.png)

### Crowded lines
![image](https://user-images.githubusercontent.com/1937582/185645074-8e7863a1-0016-4836-b4b0-6b0268f55564.png)

### Crowded bars
![image](https://user-images.githubusercontent.com/1937582/185647141-7cf1c01e-cf79-456e-a617-bd15cefb4ad3.png)

### Crowded areas
![image](https://user-images.githubusercontent.com/1937582/185647161-a8619593-ace5-4f50-8365-6030f0c1d724.png)

WiNloSt committed:
Remove non-existent route

WiNloSt committed:
Make some internal routes easier to access

WiNloSt committed:
Show data point value on static bar chart

WiNloSt committed:
Show data point values on static line chart

WiNloSt committed:
Show data point values on static area chart

WiNloSt committed:
Remove unused function

WiNloSt commented on src/metabase/pulse/render.clj:
```diff
@@ -87,6 +87,7 @@
 
         (#{:area
            :bar
+           :line
```
This is a hack that should be removed once BE is properly implemented.

WiNloSt commented on src/metabase/pulse/render/body.clj:
```diff
@@ -321,6 +321,7 @@
                 :format x-format}
      :y        {:type (or (:graph.y_axis.scale viz-settings) "linear")
                 :format y-format}
+     :show_values (boolean (:graph.show_values viz-settings))
```
This is a hack that should be removed once BE is properly implemented.

WiNloSt commented on src/metabase/pulse/render/js_svg.clj:
```diff
@@ -177,15 +177,6 @@
                                                   (json/generate-string settings)))]
     (svg-string->bytes svg-string)))
 
-(defn categorical-bar
```
This function isn't used.

adam-james-v replied:
Ah, good catch! My local dev. says that 'categorical-area' isn't used either.

WiNloSt commented on frontend/src/metabase/static-viz/components/XYChart/shapes/LineSeries.tsx:
```diff
@@ -29,17 +40,63 @@ export const LineSeries = ({
+            {showValues &&
+              s.data.map((datum, index) => {
+                // Use the similar logic as presented in https://github.com/metabase/metabase/blob/3f4ca9c70bd263a7579613971ea8d7c47b1f776e/frontend/src/metabase/visualizations/lib/chart_values.js#L130
+                const showLabelBelow =
```
Flip the data point value position using the same logic as
https://github.com/metabase/metabase/blob/3f4ca9c70bd263a7579613971ea8d7c47b1f776e/frontend/src/metabase/visualizations/lib/chart_values.js#L130-L134

#### Result
![image](https://user-images.githubusercontent.com/1937582/184912556-8f0f9741-4760-461a-9136-f582822be145.png)

WiNloSt commented on frontend/src/metabase/static-viz/components/XYChart/XYChart.tsx:
```diff
@@ -80,7 +80,7 @@ export const XYChart = ({
     settings.labels,
     style.axes.ticks.fontSize,
-    !!settings.goal,
+    !!settings.goal || !!settings.show_values,
```
normally I'd prefer `Boolean(expression)` but I use `!!` just to be consistent with the existing expression.

WiNloSt commented on frontend/src/metabase/internal/components/Layout.jsx:
```diff
@@ -106,8 +106,13 @@ export const InternalLayout = fitViewport(({ children }) => {
+          </li>
+          <li>
+            <Link className="link" to={"/_internal/static-viz"}>
+              <Label>Static Visualizations</Label>
```
Add a link to our `/_internal` route so we could access these pages more conveniently.
![image](https://user-images.githubusercontent.com/1937582/184914821-670b1bd5-0ff2-431b-b74f-9c7b4901e323.png)

WiNloSt commented on frontend/src/metabase/internal/components/Layout.jsx:
```diff
@@ -106,8 +106,13 @@ export const InternalLayout = fitViewport(({ children }) => {
             </Link>
           </li>
           <li>
-            <Link className="link" to={"/_internal/colors"}>
```
This page doesn't seem to exist anymore.

alxnddr reviewed (commented):
@WiNloSt good start! Make sure you address these cases described in the issue:
- Long values (shorten numbers like 100k, 1m)
- Crowded bars
- Multiseries bars
- Close to chart edges labels should be visible
- Negative bars
<img width="640" alt="Screen Shot 2022-08-17 at 1 24 09 PM" src="https://user-images.githubusercontent.com/14301985/185084090-80224720-9523-4171-b02c-e62774deb914.png">


<img width="638" alt="Screen Shot 2022-08-17 at 1 24 17 PM" src="https://user-images.githubusercontent.com/14301985/185084128-4fe21a1e-591d-44cf-ac9a-68c71ee281fd.png">

WiNloSt commented:
Thanks a lot @alxnddr That list looks exhaustive.

WiNloSt committed:
Improve multiple series area chart data point value legibility

WiNloSt committed:
Render stacked area chart data point values

WiNloSt committed:
Automatically format long values with compact format

adam-james-v commented:
@WiNloSt The backend changes you've made you call 'hacks'. Could you elaborate on what you mean by that? From my understanding, I think it's mostly a case of putting keys and values in weird places. For instance, pulling `:show-values` out of the viz-settings map ( `:show_values (boolean (:graph.show_values viz-settings))` ). Could you elaborate so that I know what you're looking for me to implement?

Additionally, please let me know if you plan to merge this PR and have the backend changes done on a separate PR. Thanks :)

WiNloSt commented:
@adam-james-v thanks for taking a look.

> The backend changes you've made you call 'hacks'. Could you elaborate on what you mean by that? From my understanding, I think it's mostly a case of putting keys and values in weird places. For instance, pulling `:show-values` out of the viz-settings map ( `:show_values (boolean (:graph.show_values viz-settings))` ). Could you elaborate so that I know what you're looking for me to implement?

1. `:show-values` isn't passed to static viz at all currently. That function `->ts-viz` that I modified is used by `combo-chart` which is the focus of this PR. https://github.com/metabase/metabase/blob/1383a975dca7a300cb03208f9a8a2a7a8a38d2e9/src/metabase/pulse/render/body.clj#L598
And the result of calling `->ts-viz` doesn't contain the whole viz-settings. Seems like we're picking only certain keys out of it, so that's why I added another key `:show-values`, basically I want an access to `viz-settings.graph.show_values` which records the value of this toggle
![image](https://user-images.githubusercontent.com/1937582/185542813-a1b410d5-c3ca-493e-b501-670bd0e365f4.png)
so static viz know if we want to start showing data point values for certain charts.

2. As for the added `:line` hack, currently without it, BE would determine the line chart as `:sparkline` https://github.com/metabase/metabase/blob/1383a975dca7a300cb03208f9a8a2a7a8a38d2e9/src/metabase/pulse/render.clj#L111-L117
which renders either `timelineseries-line` or `categorical-line` https://github.com/metabase/metabase/blob/1383a975dca7a300cb03208f9a8a2a7a8a38d2e9/src/metabase/pulse/render/body.clj#L725-L727
I don't really know what to deal with this, but what I wanted is for the :line chart to be rendered with `combo-chart` instead. If I understand correctly this would be done in https://github.com/metabase/metabase/issues/24671 but I need a quick way to render it now so I know my static viz implementation would work when it's integrated with BE, hence the hack

> Additionally, please let me know if you plan to merge this PR and have the backend changes done on a separate PR. Thanks :)

Yes, eventually those hacks (I now rebased them to the first 2 commits, so they're more apparent) will be removed from the PR, the reason they're there now so that I could fully test the static viz rendering with BE, to see if it works e.g. [paint-order](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/paint-order) doesn't work when rendering on the BE and I wouldn't know if I don't let BE render the chart.

And it'd be more efficient if you do it because I don't know where I should be putting things on BE that'd make sense. Also, I don't fully understand the BE work of static viz yet, though I kind of have an overview of it.

Lastly, since these are all hacks and most notably the place where I put `:show_values` if you think it might make more sense to expose it somewhere else, feels free to do so, so I could modify static viz code to read the setting from a new place accordingly. And since this is just an interface between BE <--> static viz the change would be really minimal.

WiNloSt committed:
Sample data point values on dense chart

WiNloSt committed:
Improve bar chart values legibility

WiNloSt commented on frontend/src/metabase/static-viz/components/XYChart/types.ts:
```diff
@@ -48,6 +48,7 @@ export type ChartSettings = {
     value: number;
     label: string;
   };
+  show_values?: boolean;
```
This prop could be changed later if BE decided it's best to expose `show_values` somewhere else or in any other way.

WiNloSt commented on frontend/src/metabase/static-viz/components/XYChart/utils/margin.ts:
```diff
@@ -3,7 +3,7 @@ import { CHART_PADDING } from "metabase/static-viz/components/XYChart/constants"
 
 export const LABEL_OFFSET = 10;
-export const GOAL_MARGIN = 6;
+export const MARGIN = 6;
```
just a rename to indicate that it's not only a goal line that should have margin.

WiNloSt commented on frontend/src/metabase/static-viz/components/XYChart/utils/values.ts:
```diff
@@ -0,0 +1,33 @@
+  valueProps: Partial<TextProps>,
+  chartWidth: number,
+) {
+  if (series.length === 0) {
+    return 1;
+  }
```
Added a guard check so, it won't fail when `series` is an empty array. Unlike the referenced code, combo-chart can display multiple chart at once, making `series` possible to be an empty array.

WiNloSt commented on frontend/src/metabase/static-viz/components/XYChart/utils/values.ts:
```diff
@@ -0,0 +1,33 @@
+  const totalWidth = sample.reduce(
+    (sum, sampledData) =>
+      sum +
+      measureText(
```
Use `measureText` to calculate the width instead of using `getBoundingClientRect` I haven't tested if `getBoundingClientRect` doesn't work on BE and that's why we need `measureText` in the first place, maybe @alxnddr would know?

alxnddr replied:
No, you cannot use it since there is no actual rendering of any real DOM. That would make things a lot easier if we could use `getBoundingClientRect`

WiNloSt replied:
that makes sense.

WiNloSt commented on frontend/src/metabase/static-viz/components/XYChart/utils/values.ts:
```diff
@@ -0,0 +1,33 @@
+import { measureText } from "metabase/static-viz/lib/text";
+
+// Use the same logic as in https://github.com/metabase/metabase/blob/72d23d2b8e5a1f93b288e5a8738758c9b05d3766/frontend/src/metabase/visualizations/lib/chart_values.js#L318
+export function getValueStep(
```
Almost the same logic in https://github.com/metabase/metabase/blob/72d23d2b8e5a1f93b288e5a8738758c9b05d3766/frontend/src/metabase/visualizations/lib/chart_values.js#L318

One caveat is that this is a simplified version of that logic and it doesn't render different series differently unlike the referenced logic where each series could have a different step since different series could have a different amount of data points.

This will be addressed in a followed-up PR.

WiNloSt commented on frontend/src/metabase/static-viz/lib/numbers.ts:
```diff
@@ -30,17 +31,33 @@ export const formatNumber = (number: number, options?: NumberFormatOptions) => {
-    minimumFractionDigits: decimals,
-    maximumFractionDigits: decimals != null ? decimals : 2,
-  });
+  function createFormat(compact?: boolean) {
```
I create this function inside this scope because it uses a lot of local variables, if I were to extract this outside this function it will require a lot of parameters making it less readable imo.

alxnddr commented on frontend/src/metabase/static-viz/components/LineAreaBarChart/LineAreaBarChart.tsx:
```diff
@@ -24,20 +26,20 @@ interface LineAreaBarChartProps {
   series,
   settings,
-  colors,
+  colors: instanceColors,
```
In the `StaticChart` component there is already  `getColor` function passed to the components, you just need to feed it with a palette from the Clojure side. Wdyt about sticking to it?

WiNloSt replied:
Yeah, I think that's better, I didn't think about that. Thanks for the suggestion!

alxnddr reviewed (approved):
Great job, you've handled so many different edge cases and the code looks clean!

Regarding labels outline, since it looks a lot more readable with it, you can render labels two times:
1) With the stroke and the paint-order. This one is rendered incorrectly by the BE engine so we need another one
2) Just a normal label without paint-order

<img width="588" alt="Screen Shot 2022-08-23 at 2 48 21 PM" src="https://user-images.githubusercontent.com/14301985/186140297-29d33474-d7b2-41bb-8c4c-1cf01462c953.png">

alxnddr commented:
Seems like if we taclke different series labels collisions and collisions with X axis in following PRs it will be good to go

WiNloSt commented:
@alxnddr Thanks for the idea for rendering value with stroke ❤️ that sounds more plausible.
